test/types/type_variables/ferguson/issue-16316.chpl

```diff
@@ -0,0 +1,56 @@
+config param compOption = true;
+config const option = true;
+
+proc returnsArrayType(arg) type {
+  return arg.type;
+}
+
+proc makeArray() {
+  writeln("in makeArray()");
+  var A:[1..10] int;
+  return A;
+}
+
+proc test1() {
+  writeln("test1");
+  type t = returnsArrayType(makeArray());
+}
+test1();
+
+proc test2() {
+  writeln("test2");
+  type t;
+  t = returnsArrayType(makeArray());
+}
+test2();
+
+proc test3() {
+  writeln("test3");
+  type t;
+  {
+    t = returnsArrayType(makeArray());
+  }
+}
+test3();
+
+proc test4() {
+  writeln("test4");
+  type t;
+  if compOption {
+    t = returnsArrayType(makeArray());
+  } else {
+    t = returnsArrayType(makeArray());
```

Comment:
I'm assuming it fails if we miss setting it in one of the if/else branches?

Reply:
You mean if one branch sets it and the other doesn't ? Yeah, it should be a compilation error in a case like this. We have other tests for that.